Identity proofing

pkarman opened the pull request:
**Why**

Once account has been created, to elevate the assurance level
it is necessary to proof the account.

**How**

The proofer gem brokers connections with outside vendor
services to do the proofing. There is also a 'mock' vendor
to allow for testing and UI/UX development.

**NOTE**

There are several known TODOs with this feature as a whole, including (but not limited to):

* the use of KBV
* UX around initial linking to /idv URL
* UX around soliciting user PII (one form, multiple forms, messaging, etc)
* UX around pass/fail flows for both Resolution and Confirmation
* storage of PII, both temp session and post-confirmation

However, it seems better to land this feature branch sooner rather than later, so that development can stabilize and those TODOs can be addressed from master rather than continuing from the feature branch.

sbc100 commented on app/controllers/concerns/idv_session.rb:
```diff
@@ -0,0 +1,59 @@
+  protected
+
+  def idv_question_number
+    session[:idv_question_number] ||= 0
```
From my recent investigations into use the of sessions with device, I would recommend using #user_session everywhere rather than #session.   In theory if we always sign_out correctly in should make any difference but we had issues where we only signed out the :user scope and then any variables in the top level session would persist between login sessions or different users.

Now that we have fixed that issue and we always sign_out without a limited scope i think its probably not an issue, but its probably still better/safer to store stuff in the user_session.

sbc100 commented on app/helpers/form_helper.rb:
```diff
@@ -6,4 +6,70 @@ def app_setting_value_field_for(app_setting, f)
+      ['Wisconsin', 'WI'],
+      ['Wyoming', 'WY']
+    ]
+  end
```
Is there somewhere we can pull this data from?  a gem maybe?

pkarman replied:
great question. I asked here: https://18f.slack.com/archives/ruby/p1466632124000169 yesterday and the consensus was do it yourself. I'm game for spinning this into a gem, esp since 18F/USDS would probably re-use it.

sbc100 replied:
Not a problem.  Was just checking if there was an existing solution.

sbc100 commented on app/views/idv/questions/index.html.slim:
```diff
@@ -0,0 +1,4 @@
+- if @question.choices.nil?
+  = render 'idv/questions/text'
+- else
+  = render 'idv/questions/multiple_choice'
```
Since this is pure logic would it make more sense in the controller? e.g:

```
def index
  if @question.choices.nil?
    render :question_multi_choice
  else
    render :question_text
```

pkarman replied:
good point. There seems to be different opinions on how/where to do views partials logic. I imagined that since both views actually share a lot of duplicate markup, it would make sense to refactor this into a better partial, which would mean keeping the logic in the view.

Is this a blocker for you? This and other UX changes will undoubtedly come up in a future PR.

sbc100 commented on bin/proofing-vendor-envs:
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env ruby
```
Can you add a comment here describing what this script does?

(ideally a --help would be nice too if you feel like it)

pkarman replied:
done.

sbc100 commented on spec/controllers/idv/questions_controller_spec.rb:
```diff
@@ -0,0 +1,43 @@
+  let(:resolution) { agent.start applicant }
+
+  context 'user has started proofing session' do
+    it 'requires user is signed in' do
```
Instead of checking this for every action we have been using the check for the before_filter in the controller.  (grep for  have_filters in the specs tree)

sbc100 commented:
Glad to see this landing in master!  Very exciting. 

The first line of the commit could be more descriptive, e.g. "Add initial identity proofing implementation"

pkarman commented:
thanks for the review @sbc100 -- I think I've hit all your comments.

Let me know what you think about splitting states fixture out into its own gem. I'm happy to do that, but maybe after this is merged, as a future enhancement.

sbc100 commented on app/controllers/concerns/idv_session.rb:
```diff
@@ -0,0 +1,64 @@
+
+  def idv_session
+    user_session[:idv] ||= {}
+    user_session[:idv]
```
Can you drop this line?

sbc100 commented on spec/controllers/idv/questions_controller_spec.rb:
```diff
@@ -0,0 +1,48 @@
+  end
+
+  def init_idv_session
+    sign_in(user)
```
Put this in a before block for the above scenario?

sbc100 commented:
Looks like a good initial commit to me.  Perhaps @sashax and/or @monfresh could take a pass before merging?

pkarman commented:
fyi: codeclimate still has trouble checking this branch, so will show as 🔴 

@sbc100 I've addressed your comments. @sashax worked on this PR with me and @brendansudol so maybe @monfresh should take a peek?

sashax commented:
Yeah, I was in the thick of this, and don't have any additional substantive comments to make other than thumbsup.

sbc100 commented:
@monfresh sup is up with code climate?

sbc100 commented on .rubocop.yml:
```diff
@@ -8,6 +8,7 @@ AllCops:
     - 'config/initializers/devise.rb'
     - 'db/migrate/20160405212342_initial_schema.rb'
   UseCache: true
+  TargetRubyVersion: 2.3.1
```
This is what is causing code climate to fail.. did you have to add this for this change?

pkarman replied:
ah! ok, I had added that to quiet a warning in my dev env. I can revert it.

sbc100 commented:
Great, can you squash those two changes? then lgtm

pkarman commented:
@sbc100 all 💚  :squash:
